libraries/AP_Radio/AP_Radio_bk2425.cpp

```diff
@@ -0,0 +1,1537 @@
+                value -= 4;
+        }
+        if (t_status.pps == 0)
+            value = BK_RSSI_MIN; // No packets = no RSSI
+        pwm_channels[chan-1] = value;
+        chan_count = MAX(chan_count, chan);
+    }
+
+    chan = get_pps_chan();
+    if ((chan > 0) && ((chan-1) < BEKEN_MAX_CHANNELS)) {
+        pwm_channels[chan-1] = t_status.pps; // How many packets received per second
+        chan_count = MAX(chan_count, chan);
+    }
+
+    chan = get_tx_rssi_chan();
+    if ((chan > 0) && ((chan-1) < BEKEN_MAX_CHANNELS)) {
+        pwm_channels[chan-1] = BK_RSSI_DEFAULT; // Fixed value that will not update (halfway in the RSSI range for Cypress chips, 0..31)
+        chan_count = MAX(chan_count, chan);
+    }
+
+    chan = get_tx_pps_chan();
+    if ((chan > 0) && ((chan-1) < BEKEN_MAX_CHANNELS)) {
+        pwm_channels[chan-1] = tx_pps;
+        chan_count = MAX(chan_count, chan);
+    }
+    
+    // Every second, update the statistics
+    if (now - last_pps_ms > 1000) {
+        last_pps_ms = now;
+        t_status.pps = stats.recv_packets - last_stats.recv_packets;
+        last_stats = stats;
+        if (stats.lost_packets != 0 || stats.timeouts != 0) {
+            Debug(3,"lost=%lu timeouts=%lu\n", stats.lost_packets, stats.timeouts);
+        }
+        stats.lost_packets=0;
+        stats.timeouts=0;
+        if (have_tx_pps == 1) // Have we had tx pps recently?
+            tx_pps = 0;
+        if (have_tx_pps == 2) // We have had it at some time
+            have_tx_pps = 1; // Not recently
+    }
+    return chan_count;
+}
+
+/*
+  return time of last receive in microseconds
+ */
+uint32_t AP_Radio_beken::last_recv_us(void)
+{
+    return synctm.packet_timer;
+}
+
+/*
+  send len bytes as a single packet
+ */
+bool AP_Radio_beken::send(const uint8_t *pkt, uint16_t len)
+{
+    // disabled for now
+    return false;
+}
+
+// Borrow the CRC32 algorithm from AP_HAL_SITL
+// Not exactly fast algorithm as it is bit based
+#define CRC32_POLYNOMIAL 0xEDB88320L
+static uint32_t CRC32Value(uint32_t icrc)
+{
+    int i;
+    uint32_t crc = icrc;
+    for ( i = 8 ; i > 0; i-- )
+    {
+        if ( crc & 1 )
+            crc = ( crc >> 1 ) ^ CRC32_POLYNOMIAL;
+        else
+            crc >>= 1;
+    }
+    return crc;
+}
+
+static uint32_t CalculateBlockCRC32(uint32_t length, const uint8_t *buffer, uint32_t crc)
+{
+    while ( length-- != 0 )
+    {
+        crc = ((crc >> 8) & 0x00FFFFFFL) ^ (CRC32Value(((uint32_t) crc ^ *buffer++) & 0xff));
+    }
+    return( crc );
+}
+
+/*
+  initialise the radio
+ */
+void AP_Radio_beken::radio_init(void)
+{
+    DebugPrintf(1, "radio_init\r\n");
+    beken.SetRBank(1);
+    uint8_t id = beken.ReadReg(BK2425_R1_WHOAMI); // id is now 99
+    beken.SetRBank(0); // Reset to default register bank.
+
+    if (id != BK_CHIP_ID_BK2425) {
+        
+        Debug(1, "bk2425: radio not found\n"); // We have to keep trying  because it takes time to initialise
+        return; // Failure
+    }
+
+	{
```

Comment:
I guess this extra set of brackets is here to reduce the RAM usage in case this radio_init function exits early?

Reply:
just to keep min scope on vars